### PR TITLE
feat: Implement PowerShell support

### LINKED ISSUE #126: Add prompt support for PowerShell
## Feature Request
Support for [PowerShell](https://github.com/PowerShell/PowerShell) would be extra cool!

I'd be more than happy to put together a pull request for this.

#### Is your feature request related to a problem? Please describe.
Cannot use Starship if workflow is mainly with PowerShell

#### Describe the solution you'd like
Add support for PowerShell for the Starship prompt

#### Describe alternatives you've considered
Using zsh/fish via [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/about), and invoking powershell scripts from there. But obviously not as nice as having it in PS

### PR DESCRIPTION
**Pull request by davidkna**
#### Description

Now that #246 is closed I've taken another stab at implementing support for powershell.
Powershell is a cross-platform shell made by Microsoft.

The prompt loads and everything seems to work fine so far. If there's anything specific I should look out for please tell me.

#### Motivation and Context

Closes #126

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [ ] I have tested using **Linux**
- [x] I have tested using **Windows**
I haven't been able to test my battery status change yet, because I don't have a device handy for that.

#### Checklist:

- [x] I have updated the documentation accordingly.
I believe I changed the necessary parts. The Japanese part still needs updating.
- [ ] I have updated the tests accordingly.
I haven't found any test infrastructure for specific shells thus I haven't added any tests yet.

### CODE REVIEW

**Comment by matchai** [comment]
Before this can be merged, we need to have Windows added to the test and build processes in CI:
https://github.com/starship/starship/blob/master/.github/workflows/continuous-integration.yml#L46
https://github.com/starship/starship/blob/master/azure-pipelines.yml#L22-L24

**Comment by davidkna** [comment]
It looks like I don't have permission to push to `.github/workflows/continuous-integration.yml` but those tests should just work with `windows-latest` added to the os matrix.
As for binary releases for windows: Is the existing code working and just needs uncommenting or does it need any fix-ups?

**Comment by matchai** [comment]
Last we tried, many of the existing tests were failing. I think some of our test utilities will need to be updated to be Windows compatible.
I'll push the CI update to your branch when I'm at my computer.

**Comment by davidkna** [comment]
I think the issue is that python.org [doesn't have a binary windows release for 3.6.9](https://www.python.org/downloads/release/python-369/)  (anymore?).
Would you mind changing the python version to `3.6`, `3.6.8`, `3.7` or `3.7.4`?

**Comment by davidkna** [comment]
I can't resolve the conflicts because now I don't have write access to the branch anymore. 😄

**Comment by chipbuster** [comment]
> I can't resolve the conflicts because now I don't have write access to the branch anymore. 

Wait, is that supposed to happen? It doesn't seem like that's supposed to happen.

**Comment by davidkna** [comment]
wait a second - i think i found the issue

**Comment by matchai** [comment]
It doesn't seem like the updated workflow file is being used in CI. 😕

**Comment by matchai** [comment]
Ah... forgot to push the CI changes. 🙃

File: tests/testsuite/git_status.rs [hunk 1]
```diff
@@ -5,6 +5,14 @@ use std::process::Command;
 
 use crate::common::{self, TestCommand};
 
+/// Ensures commands finish on windows
+#[cfg(not(windows))]
+pub fn barrier() {}
+#[cfg(windows)]
+pub fn barrier() {
+    std::thread::sleep(std::time::Duration::from_millis(500));
```
Comment (Southclaws) [inline tests/testsuite/git_status.rs]: This feels like a hack, why does it work and what is it solving?

Also, regarding the comment, is it _Windows_ specific or is it _PowerShell_ specific?

Comment (nickwb) [reply tests/testsuite/git_status.rs]: Agree, I would like to understand this better.

Comment (davidkna) [reply tests/testsuite/git_status.rs]: I looks like right after the calls to git the filesystem state hasn't finished updating yet causing some of the tests to fail ([test run before change](https://github.com/starship/starship/runs/249566711), [test run after change](https://github.com/starship/starship/runs/249783836)). I also added these barriers after each call to git, even where it hadn't caused issues yet, to reduce possible test flakiness. I believe it isn't trivial to start a full fs sync on windows.
This addition is Windows specific though other operating systems could have the same issue. None of the tests cover anything powershell-specific yet.

**Comment by davidkna** [comment]
The `aws` and `conda` tests pass when the project is build with `-Z unstable-option` beforehand.

File: tests/testsuite/git_status.rs (hunk 1 above)
Comment (Southclaws) [reply tests/testsuite/git_status.rs]: Ah okay, good to know thanks!

**Comment by davidkna** [comment]
Not sure what's wrong with the failing git tests. Can't reproduce locally.

File: src/init/starship.ps1 [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env pwsh
+
+# Starship assumes UTF-8
+[Console]::OutputEncoding = [System.Text.Encoding]::UTF8
+function global:prompt {
+    $out = $null
+    $jobs = (Get-Job | Where-Object { $_.State -eq 'Running' }).Count
```
Comment (nickcox) [inline src/init/starship.ps1]: Unsolicited (sorry!) but FWIW, this breaks on my mac when `Get-Job | Where-Object { $_.State -eq 'Running' }` returns one or fewer results.

```Suggestion
$jobs = if ($j = Get-Job | Where-Object { $_.State -eq 'Running' }) { $j.Count }
```
or

```Suggestion
Get-Job | Where-Object { $_.State -eq 'Running' } | Measure-Object | Select-Object -Expand Count
```

Comment (davidkna) [reply src/init/starship.ps1]: Please try again with the newest commit

**Comment by davidkna** [comment]
All tests green!

File: tests/testsuite/common.rs [hunk 3]
```diff
@@ -8,9 +8,15 @@ use std::{env, fs, io, process};
 static MANIFEST_DIR: Lazy<&'static Path> = Lazy::new(|| Path::new(env!("CARGO_MANIFEST_DIR")));
 static EMPTY_CONFIG: Lazy<PathBuf> = Lazy::new(|| MANIFEST_DIR.join("empty_config.toml"));
 
+#[cfg(windows)]
```
Comment (matchai) [inline tests/testsuite/common.rs]: Nice 👍

File: docs/README.md [hunk 4]
```diff
@@ -97,3 +97,12 @@ footer: ISC Licensed | Copyright © 2019-present Starship Contributors
 
    eval "$(starship init zsh)"
    ```
+
+   #### Powershell
```
Comment (matchai) [inline docs/README.md]: Solid docs 👍

File: src/init/mod.rs [hunk 5]
```diff
@@ -94,6 +94,13 @@ fi"#,
             );
             Some(script)
         }
+        Some("powershell") => {
+            let script = format!(
+                "Invoke-Expression (@(&\"{}\" init powershell --print-full-init) -join \"`n\")",
```
Comment (matchai) [inline src/init/mod.rs]: I'm not a Powershell user, so I'm not super familiar with what's going on here, specifically the symbols (`(@(&`), nor why ``-join \"`n\"`` is needed.

Would the average Powershell user understand this, or should comments be added?

File: src/init/starship.ps1 [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+    if ($lastCmd = Get-History -Count 1) {
+        $duration = [math]::Round(($lastCmd.EndExecutionTime - $lastCmd.StartExecutionTime).TotalSeconds)
+        $out = @(&::STARSHIP:: prompt --status=$lastexitcode --jobs=$jobs --cmd-duration=$duration)
+    }
+    else {
+        $out = @(&::STARSHIP:: prompt --status=$lastexitcode --jobs=$jobs)
+    }
+
+    $out -join "`n"
```
Comment (matchai) [inline src/init/starship.ps1]: I understand everything in this file except for this line.
What does ``$out -join "`n"`` do here?

File: src/main.rs [hunk 7]
```diff
@@ -34,7 +34,7 @@ fn main() {
     let shell_arg = Arg::with_name("shell")
         .value_name("SHELL")
         .help(
-            "The name of the currently running shell\nCurrently supported options: bash, zsh, fish",
+            "The name of the currently running shell\nCurrently supported options: bash, zsh, fish, powershell",
```
Comment (matchai) [inline src/main.rs]: 👍

File: tests/testsuite/common.rs [hunk 8]
```diff
@@ -38,9 +44,13 @@ pub fn render_module(module_name: &str) -> process::Command {
 
 /// Create a temporary directory with full access permissions (rwxrwxrwt).
 pub fn new_tempdir() -> io::Result<tempfile::TempDir> {
-    //  Using `tempfile::TempDir` directly creates files on macOS within
-    // "/var/folders", which provides us with restricted permissions (rwxr-xr-x)
-    tempfile::tempdir_in("/tmp")
+    if cfg!(not(windows)) {
+        //  Using `tempfile::TempDir` directly creates files on macOS within
+        // "/var/folders", which provides us with restricted permissions (rwxr-xr-x)
+        tempfile::tempdir_in("/tmp")
```
Comment (matchai) [inline tests/testsuite/common.rs]: Running some tests on another branch, I think we may no longer need this for tests to pass. Might've been solved by an update in `tempfile` or `docker`. Not too sure 🤷‍♂

File: tests/testsuite/nodejs.rs [hunk 9]
```diff
@@ -4,11 +4,28 @@ use std::io;
 
 use crate::common;
 
+fn render_node_module() -> std::process::Command {
+    let mut command = common::render_module("nodejs");
+    if cfg!(windows) {
```
Comment (matchai) [inline tests/testsuite/nodejs.rs]: Could we get a comment here explaining why `SYSTEMROOT` is needed in Windows?

File: tests/testsuite/common.rs (hunk 8 above)
Comment (davidkna) [reply tests/testsuite/common.rs]: If it's still needed I think this should be changed to `cfg!(darwin)`

File: src/init/mod.rs (hunk 5 above)
Comment (davidkna) [reply src/init/mod.rs]: I've added explanations to this module and the powershell script.

File: tests/testsuite/nodejs.rs (hunk 9 above)
Comment (davidkna) [reply tests/testsuite/nodejs.rs]: Done!

**Comment by matchai** [comment]
Thank you for tackling this, @davidkna! I'm very excited to have starship officially support Powershell and Windows! 🎉 

@all-contributors Please add @davidkna for code, docs, tests.

**Pull request merged**